tuplex/core/src/TraceVisitor.cc

```diff
@@ -944,4 +969,277 @@ namespace tuplex {
 
     }
 
+    void TraceVisitor::visit(NFor* node) {
+        assert(node && node->target && node->expression && node->suite_body);
+        setLastParent(node);
+
+        // expression type check
+        auto exprType = node->expression->getInferredType();
+        if (!(exprType.isListType() || exprType.isTupleType() || exprType == python::Type::STRING ||
+              exprType == python::Type::RANGE || exprType.isIteratorType())) {
+            throw std::runtime_error("unsupported for loop expression type encountered in TraceVisitor.cc");
```

Comment:
does this result in marking the function as uncompilable higher up the framework chain? I.e., is fallback correctly invoked?

Reply:
Added new CompileError TYPE_ERROR_UNSUPPORTED_LOOP_EXPRESSION_TYPE. Added logic to check and add this CompileError through TypeAnnotatorVisitor, so that unsupported loops don't end up in TraceVisitor. The check in TraceVisitor is still kept.
Now unsupported for loop expression will invoke fallback.